src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +116,256 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+                    memory.Clear();
+                    MarkTransitiveOrigin(directPkg, directPkg.PackageIdentity, targetFxGraph, memory, key);
+                }
+            }
+
+            IsTransitiveComputationNeeded = false;
+
+            return GetCachedTransitiveOrigin(transitivePackage);
+        }
+
+        /// <summary>
+        /// Returns <see cref="PackageSpec"/> found in assets file (project.assets.json)
+        /// </summary>
+        /// <param name="token">Cancellation token</param>
+        /// <returns>An <see cref="RestoreGraphRead"/> object</returns>
+        /// <remarks>Projects need to be NuGet-restored before calling this function</remarks>
+        internal async Task<RestoreGraphRead> GetCachedPackageSpecAsync(CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            string assetsFilePath = await GetAssetsFilePathAsync();
+            var assets = new FileInfo(assetsFilePath);
+
+            PackageSpec currentPackageSpec = await GetPackageSpecAsync(token);
+            PackageSpec cachedPackageSpec = null;
+            bool cacheHitPackageSpec = _lastPackageSpec != null && _lastPackageSpec.TryGetTarget(out cachedPackageSpec);
+
+            bool cacheMissAssets = (assets.Exists && assets.LastWriteTimeUtc > _lastTimeAssetsModified);
+            bool isCacheHit = false;
+
+            if (cacheMissAssets || IsPackageSpecDifferent(currentPackageSpec, cachedPackageSpec))
+            {
+                _lastTimeAssetsModified = assets.LastWriteTimeUtc;
+                _lastPackageSpec = new WeakReference<PackageSpec>(currentPackageSpec);
+                IsInstalledAndTransitiveComputationNeeded = true;
+                IsTransitiveComputationNeeded = true;
+            }
+            else if (cacheHitPackageSpec && cachedPackageSpec != null)
+            {
+                isCacheHit = true;
+            }
+
+            return new RestoreGraphRead(currentPackageSpec, isCacheHit);
+        }
+
+        /// <summary>
+        /// Obtains targets section from project assets file (project.assets.json)
+        /// </summary>
+        /// <param name="ct">Cancellation token for async operation</param>
+        /// <returns>A lis of dependencies, indexed by framework/RID</returns>
+        /// <remarks>Assets file reading occurs in a background thread</remarks>
+        /// <seealso cref="GetAssetsFilePathAsync"/>
+        protected async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            await TaskScheduler.Default;
+
+            string assetsFilePath = await GetAssetsFilePathAsync();
+            LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
+
+            return lockFile?.Targets;
+        }
+
+        /// <summary>
+        /// Runs Depth First Search recursively to mark current and dependend nodes with top dependency
```

Comment:
out of curiosity, why did we choose DFS over BFS? I could be wrong, but my understanding is BFS will find the required transitive package sooner because we are searching for shortest path to the transitive package from the direct dependency.

Reply:
Yes, we use DFS because we don't want shortest path to an installed package. We want all packages that install a transitive dependence.